src/common/meta/src/kv_backend/etcd.rs

```diff
@@ -33,55 +33,163 @@ use crate::rpc::store::{
     }
+    txn_res
 }
 
-impl<'a> From<KvPair<'a>> for KeyValue {
-    fn from(kv: KvPair<'a>) -> Self {
-        Self {
-            key: kv.0.key().to_vec(),
-            value: kv.0.value().to_vec(),
+fn key_prepend_root(root: &[u8], mut key: Vec<u8>) -> Vec<u8> {
+    let mut new_key = root.to_vec();
+    new_key.append(&mut key);
+    new_key
+}
+
+// see namespace.prefixInterval - https://github.com/etcd-io/etcd/blob/v3.5.10/client/v3/namespace/util.go
+fn range_end_prepend_root(root: &[u8], mut range_end: Vec<u8>) -> Vec<u8> {
+    if range_end == [0] {
+        // the edge of the keyspace
+        let mut new_end = root.to_vec();
+        let mut ok = false;
+        for i in (0..new_end.len()).rev() {
+            new_end[i] = new_end[i].wrapping_add(1);
+            if new_end[i] != 0 {
+                ok = true;
+                break;
+            }
         }
+        if !ok {
+            // 0xff..ff => 0x00
+            new_end = vec![0];
+        }
+        new_end
+    } else if !range_end.is_empty() {
+        let mut new_end = root.to_vec();
+        new_end.append(&mut range_end);
+        new_end
+    } else {
+        vec![]
     }
 }
 
-// Maximum number of operations permitted in a transaction.
-// The etcd default configuration's `--max-txn-ops` is 128.
-//
-// For more detail, see: https://etcd.io/docs/v3.5/op-guide/configuration/
-const MAX_TXN_SIZE: usize = 128;
+fn txn_prepend_root(root: &[u8], mut txn: KvTxn) -> KvTxn {
+    use super::txn::TxnOp as KvTxnOp;
+
+    fn op_prepend_root(root: &[u8], op: KvTxnOp) -> KvTxnOp {
+        match op {
+            KvTxnOp::Put(k, v) => KvTxnOp::Put(key_prepend_root(root, k), v),
+            KvTxnOp::Get(k) => KvTxnOp::Get(key_prepend_root(root, k)),
+            KvTxnOp::Delete(k) => KvTxnOp::Delete(key_prepend_root(root, k)),
+        }
+    }
+
+    txn.req.success = txn
+        .req
+        .success
+        .drain(..)
+        .map(|op| op_prepend_root(root, op))
+        .collect();
+
+    txn.req.failure = txn
+        .req
+        .failure
+        .drain(..)
+        .map(|op| op_prepend_root(root, op))
+        .collect();
+
+    txn.req.compare = txn
+        .req
+        .compare
+        .drain(..)
+        .map(|cmp| super::txn::Compare {
+            key: key_prepend_root(root, cmp.key),
+            cmp: cmp.cmp,
+            target: cmp.target,
+        })
+        .collect();
+
+    txn
+}
 
 pub struct EtcdStore {
+    root: Vec<u8>,
     client: Client,
 }
 
 impl EtcdStore {
-    pub async fn with_endpoints<E, S>(endpoints: S) -> Result<KvBackendRef>
+    pub async fn with_endpoints<R, E, S>(root: R, endpoints: S) -> Result<KvBackendRef>
     where
+        R: Into<Vec<u8>>,
         E: AsRef<str>,
         S: AsRef<[E]>,
     {
         let client = Client::connect(endpoints, None)
             .await
             .context(error::ConnectEtcdSnafu)?;
 
-        Ok(Self::with_etcd_client(client))
+        Ok(Self::with_etcd_client(root, client))
```

Comment:
How about just making this feature available for test purposes? And remain no change of original implementation.

Reply:
Updated.